the-collab-lab/tcl-10-smart-shopping-list
# Issue #10 Complete

## Description



### What does this code change?

When new purchase is added update purchase frequency in Firestore. Purchase frequency is predicted using standard deviation of intervals between purchase dates.

### Why did I choose this approach?

We wanted to update the existing estimates.js function using statistical models in order to provide a more accurate result. We also thought it would be fun to dive into creating our own algorithm and work with built-in javascript Math functions.

### Did I learn anything worth sharing?

- Getting more experience sequentially executing Firebase/Firestore methods (specifically chaining firebase/firestore functions).
- Using built in javascript Math functions.

## Related Issue

closes #10 

## Acceptance Criteria

When a purchase is recorded, the estimated number of days until the next purchase date should be calculated and recorded in the database

## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

## Testing Steps / QA Criteria

1. Use token `film stag lipid`
2. Go to Firestore database and make sure index 6 in your assigned grocery item's purchase dates array is the number 10.
2. Click your grocery item in app
3. Item in index 6 will be ignored because it's an outlier and has been removed from our frequency calculation.
4. Updated frequency will be approximately the same number in the database


`debugger`  was an invaluable tool when stepping through our calculateFrequency function. If you want to see how we calculate the new frequency in the `calculateFrequency` function we highly recommend it. Just insert `debugger;` at the beginning of the `calculateFrequency` function. 


### Here are some online calculators we used to test our functions:

Standard Deviation: 
https://www.calculator.net/standard-deviation-calculator.html

Mean:
https://www.calculatorsoup.com/calculators/statistics/mean-median-mode.php

Z-Index:
https://www.calculatorsoup.com/calculators/statistics/z-score-calculator.php
